cloudcompose/ecs/controller.py

```diff
@@ -0,0 +1,299 @@
+
+    def cleanup(self):
+        """
+        Remove launch configs and autoscaling group
+        """
+        cloud_controller = CloudController(self.cloud_config)
+        cloud_controller.cleanup()
+
+    def cluster_health(self):
+        """
+        ECS cluster must be active, EC2 instances must be active, and services must be active.
+        :return: boolean representing health of entire ECS cluster
+        """
+        health_checks = [
+            self._cluster_health(),
+            self._instance_health(),
+            self._service_health()
+        ]
+        return all(health_checks)
+
+    def upgrade(self, single_step):
+        """
+        Replaces existing ECS container instances
+        :param single_step: Whether to execute a single step (defaults to entire workflow)
+        :return: None
+        """
+        servers = self._get_servers()
+        workflow = UpgradeWorkflow(self, self.config_data['name'], servers)
+
+        # Start upgrading container instances
+        if single_step:
+            print("Running single step")
+            workflow.step()
+        else:
+            print("Starting upgrade of container instances:")
+            while workflow.step():
+                sleep(10)
+
+    def instance_status(self, instance_id):
+        filters = [{ 'Name': 'instance-id', 'Values': [instance_id] }]
+        instances = self._ec2_describe_instances(Filters=filters)['Reservations']
+        if len(instances) != 1:
+            raise Exception('Expected one instance for %s and got %s' % (instance_id, len(instances)))
+        return instances[0]['Instances'][0]['State']['Name']
+
+    def _get_servers(self):
+        """
+        Describe instances for UpgradeWorkflow
+        """
+        instance_ids = [server['ec2InstanceId'] for server in self._get_ecs_instances()]
+        filters = [{'Name': 'instance-state-name', 'Values': ['running']}]
+
+        describe_instances = self._ec2_describe_instances(InstanceIds=instance_ids,
+                                                          Filters=filters)
+        servers = list(chain.from_iterable([
+            reservation.get('Instances', []) for reservation in describe_instances.get('Reservations', [])
+        ]))
+
+        return [Server(private_ip=server['PrivateIpAddress'],
+                instance_id=server['InstanceId'],
+                instance_name=self.name) for server in servers]
+
+    def _get_cluster(self):
+        try:
+            clusters = self._ecs_describe_clusters(clusters=[self.name, ])
+            if not clusters['clusters']:
+                raise CloudComposeException("{} cluster could not be found".format(self.name))
+            return clusters['clusters']
+        except KeyError:
+            raise CloudComposeException("Could not retrieve cluster status for {}".format(self.name))
+
+    def _get_ecs_services(self):
+        cluster_services = self._ecs_list_services(cluster=self.name)
+        try:
```

Comment:
I feel like this is syntactically confusing.  I think you could achieve the same goal by doing:

```python
cluster_services.get("serviceArns")
if cluster_services is not None:
    describe_services = self._ecs_describe_services(cluster=self.name,
                                                 services=cluster_services['serviceArns'])
    return describe_services['services']

raise CloudComposeException("Services could not be retrieved for {}".format(self.name))
```